File: src/lib/api/skin.ts
/** API domain: appearance skin packs / presets / catalog */

import { invoke, listen } from "./host";
import type { SkinPackPreview } from "../skinPack";
import type { CatalogPack, SkinCatalogSource } from "../skinCatalog";
import type { PendingSkinImport } from "../skinImportPending";

export type SkinPresetListItem = {
  id: string;
  sourceId: string;
  name: string;
  description: string;
  author: string;
  createdAt: number;
  updatedAt: number;
  skin: string;
  scrim: number;
  hasWallpaper: boolean;
  kind?: string;
  bytes: number;
  previewRel?: string;
};

export type SkinPresetListResult = {
  presets: SkinPresetListItem[];
  usage: { bytes: number; budget: number; hasUndo: boolean };
};

export async function skinPickOpen(): Promise<string | null> {
  return invoke<string | null>("skin_pick_open");
}

export async function skinPickSave(defaultName?: string): Promise<string | null> {
  return invoke<string | null>("skin_pick_save", { defaultName: defaultName ?? null });
}

export async function skinPackInspect(path: string): Promise<SkinPackPreview> {
  return invoke<SkinPackPreview>("skin_pack_inspect", { path });
}

export async function skinInspectAbort(inspectId: string): Promise<void> {
  await invoke<void>("skin_inspect_abort", { inspectId });
}

export type SkinExportResult = {
  warning?: "ffmpeg_unavailable" | string | null;
};

export async function skinPackExport(
  destPath: string,
  stagingId: string | null,
  manifest: unknown,
): Promise<SkinExportResult> {
  return invoke<SkinExportResult>("skin_pack_export", {
    destPath,
    stagingId,
    manifest,
  });
}

export async function skinStagingBegin(): Promise<{ uploadId: string }> {
  return invoke<{ uploadId: string }>("skin_staging_begin");
}

export async function skinStagingAppend(
  stagingId: string,
  chunkBase64: string,
): Promise<number> {
  return invoke<number>("skin_staging_append", { stagingId, chunkBase64 });
}

export async function skinStagingAbort(stagingId: string): Promise<void> {
  await invoke<void>("skin_staging_abort", { stagingId });
}

export async function skinPresetList(): Promise<SkinPresetListResult> {
  return invoke<SkinPresetListResult>("skin_preset_list");
}

export async function skinPresetSaveFromUpload(
  stagingId: string,
  manifest: unknown,
): Promise<SkinPresetListItem> {
  return invoke<SkinPresetListItem>("skin_preset_save_from_upload", {
    stagingId,
    manifest,
  });
}

export async function skinPresetSaveFromInspect(
  inspectId: string,
): Promise<SkinPresetListItem> {
  return invoke<SkinPresetListItem>("skin_preset_save_from_inspect", {
    inspectId,
  });
}

export async function skinPresetMaterialize(id: string): Promise<SkinPackPreview> {
  return invoke<SkinPackPreview>("skin_preset_materialize", { id });
}

export async function skinPresetDelete(id: string): Promise<void> {
  await invoke<void>("skin_preset_delete", { id });
}

export async function skinPresetRename(
  id: string,
  name: string,
): Promise<SkinPresetListItem> {
  return invoke<SkinPresetListItem>("skin_preset_rename", { id, name });
}

export async function skinPresetReplaceFromUpload(
  id: string,
  stagingId: string,
  manifest: unknown,
): Promise<SkinPresetListItem> {
  return invoke<SkinPresetListItem>("skin_preset_replace_from_upload", {
    id,
    stagingId,
    manifest,
  });
}

export async function skinPresetExport(
  id: string,
  destPath: string,
): Promise<SkinExportResult> {
  return invoke<SkinExportResult>("skin_preset_export", { id, destPath });
}

export async function skinUndoPrepare(): Promise<string> {
  return invoke<string>("skin_undo_prepare");
}

export async function skinUndoAppend(
  snapshotId: string,
  chunkBase64: string,
): Promise<number> {
  return invoke<number>("skin_undo_append", { snapshotId, chunkBase64 });
}

export async function skinUndoCommit(
  snapshotId: string,
  manifest: unknown,
): Promise<void> {
  await invoke<void>("skin_undo_commit", { snapshotId, manifest });
}

export async function skinUndoAbort(snapshotId: string): Promise<void> {
  await invoke<void>("skin_undo_abort", { snapshotId });
}

export async function skinCatalogFetch(
  sourceId: string,
  force?: boolean,
): Promise<CatalogPack[]> {
  return invoke<CatalogPack[]>("skin_catalog_fetch", {
    sourceId,
    force: force ?? false,
  });
}

export async function skinCatalogDownload(
  sourceId: string,
  packId: string,
): Promise<SkinPackPreview> {
  return invoke<SkinPackPreview>("skin_catalog_download", { sourceId, packId });
}

export async function skinCatalogPreviewPath(
  sourceId: string,
  packId: string,
): Promise<string | null> {
  return invoke<string | null>("skin_catalog_preview_path", { sourceId, packId });
}

export async function skinSourcesList(): Promise<SkinCatalogSource[]> {
  return invoke<SkinCatalogSource[]>("skin_sources_list");
}

export async function skinSourcesAdd(
  url: string,
  label?: string,
): Promise<SkinCatalogSource> {
  return invoke<SkinCatalogSource>("skin_sources_add", { url, label: label ?? "" });
}

export async function skinSourcesRemove(id: string): Promise<void> {
  await invoke<void>("skin_sources_remove", { id });
}

export async function skinSourcesSetEnabled(
  id: string,
  enabled: boolean,
): Promise<SkinCatalogSource> {
  return invoke<SkinCatalogSource>("skin_sources_set_enabled", { id, enabled });
}

export async function skinImportTakePending(): Promise<PendingSkinImport | null> {
  return invoke<PendingSkinImport | null>("skin_import_take_pending");
}

export async function skinPackFetchUrl(href: string): Promise<SkinPackPreview> {
  return invoke<SkinPackPreview>("skin_pack_fetch_url", { href });
}

export function listenSkinImportPending(handler: () => void): Promise<() => void> {
  return listen<unknown>("skin://import-pending", () => handler());
}

File: src-tauri/src/commands/skin.rs
// Skin pack / preset / catalog / deeplink command wrappers (included into commands).

use std::path::PathBuf;

use crate::path_scope;
use crate::skin_catalog;
use crate::skin_deeplink::{self, PendingSkinImport, PendingSlot};
use crate::skin_pack::{self, SkinPackPreviewDto};
use crate::skin_presets::{self, PresetIndexEntry};
use crate::skin_staging;

fn desktop_only() -> Result<(), String> {
    Ok(())
}

#[tauri::command]
pub async fn skin_pick_open() -> Result<Option<String>, String> {
    desktop_only()?;
    let file = tauri::async_runtime::spawn_blocking(|| {
        rfd::FileDialog::new()
            .set_title("Open appearance pack")
            .add_filter("Grok skin pack", &["grokskin", "zip"])
            .pick_file()
    })
    .await
    .map_err(|e| e.to_string())?;
    Ok(file.map(|p| {
        path_scope::grant_path(&p);
        p.display().to_string()
    }))
}

#[tauri::command]
pub async fn skin_pick_save(default_name: Option<String>) -> Result<Option<String>, String> {
    desktop_only()?;
    let name = default_name.unwrap_or_else(|| "skin.grokskin".into());
    let file = tauri::async_runtime::spawn_blocking(move || {
        rfd::FileDialog::new()
            .set_title("Save appearance pack")
            .set_file_name(&name)
            .add_filter("Grok skin pack", &["grokskin"])
            .save_file()
    })
    .await
    .map_err(|e| e.to_string())?;
    Ok(file.map(|p| {
        path_scope::grant_path(&p);
        p.display().to_string()
    }))
}

#[tauri::command]
pub async fn skin_pack_inspect(path: String) -> Result<SkinPackPreviewDto, String> {
    let p = PathBuf::from(path);
    path_scope::grant_path(&p);
    tauri::async_runtime::spawn_blocking(move || skin_pack::inspect_pack(&p, "file"))
        .await
        .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn skin_inspect_abort(inspect_id: String) -> Result<(), String> {
    skin_staging::abort_inspect(&inspect_id)
}

#[tauri::command]
pub async fn skin_pack_export(
    dest_path: String,
    staging_id: Option<String>,
    manifest: serde_json::Value,
) -> Result<crate::skin_pack::SkinExportResult, String> {
    let dest = PathBuf::from(dest_path);
    path_scope::grant_path(&dest);
    tauri::async_runtime::spawn_blocking(move || {
        let wallpaper = if let Some(id) = staging_id.as_deref() {
            let dir = crate::skin_staging::consume_upload(id)?;
            let blob = dir.join("blob.bin");
            let ext = manifest
                .get("wallpaper")
                .and_then(|w| w.get("file"))
                .and_then(|f| f.as_str())
                .and_then(|f| f.rsplit('.').next())
                .unwrap_or("bin");
            let named = dir.join(format!("wallpaper.{ext}"));
            if blob.is_file() {
                let _ = std::fs::copy(&blob, &named);
                Some(named)
            } else {
                None
            }
        } else {
            None
        };
        let r = skin_pack::export_pack(&dest, &manifest, wallpaper.as_deref());
        if let Some(id) = staging_id.as_deref() {
            let _ = crate::skin_staging::abort_upload(id);
        }
        r
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn skin_staging_begin() -> Result<skin_staging::StagingBegin, String> {
    skin_staging::begin_upload()
}

#[tauri::command]
pub async fn skin_staging_append(staging_id: String, chunk_base64: String) -> Result<u64, String> {
    tauri::async_runtime::spawn_blocking(move || {
        skin_staging::append_upload(&staging_id, &chunk_base64)
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn skin_staging_abort(staging_id: String) -> Result<(), String> {
    skin_staging::abort_upload(&staging_id)
}

#[tauri::command]
pub async fn skin_preset_list() -> Result<serde_json::Value, String> {
    let presets = skin_presets::list_presets()?;
    Ok(serde_json::json!({
        "presets": presets,
        "usage": skin_presets::disk_usage(),
    }))
}

#[tauri::command]
pub async fn skin_preset_save_from_upload(
    staging_id: String,
    manifest: serde_json::Value,
) -> Result<PresetIndexEntry, String> {
    tauri::async_runtime::spawn_blocking(move || {
        skin_presets::save_from_upload(&staging_id, manifest)
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn skin_preset_save_from_inspect(inspect_id: String) -> Result<PresetIndexEntry, String> {
    tauri::async_runtime::spawn_blocking(move || skin_presets::save_from_inspect(&inspect_id))
        .await
        .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn skin_preset_materialize(id: String) -> Result<SkinPackPreviewDto, String> {
    tauri::async_runtime::spawn_blocking(move || skin_presets::materialize(&id))
        .await
        .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn skin_preset_delete(id: String) -> Result<(), String> {
    skin_presets::delete_preset(&id)
}

#[tauri::command]
pub async fn skin_preset_rename(id: String, name: String) -> Result<PresetIndexEntry, String> {
    skin_presets::rename_preset(&id, &name)
}

#[tauri::command]
pub async fn skin_preset_replace_from_upload(
    id: String,
    staging_id: String,
    manifest: serde_json::Value,
) -> Result<PresetIndexEntry, String> {
    tauri::async_runtime::spawn_blocking(move || {
        skin_presets::replace_from_upload(&id, &staging_id, manifest)
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn skin_preset_export(
    id: String,
    dest_path: String,
) -> Result<crate::skin_pack::SkinExportResult, String> {
    let dest = PathBuf::from(dest_path);
    path_scope::grant_path(&dest);
    tauri::async_runtime::spawn_blocking(move || skin_presets::export_preset(&id, &dest))
        .await
        .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn skin_undo_prepare() -> Result<String, String> {
    skin_presets::undo_prepare()
}

#[tauri::command]
pub async fn skin_undo_append(snapshot_id: String, chunk_base64: String) -> Result<u64, String> {
    tauri::async_runtime::spawn_blocking(move || skin_presets::undo_append(&snapshot_id, &chunk_base64))
        .await
        .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn skin_undo_commit(snapshot_id: String, manifest: serde_json::Value) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || skin_presets::undo_commit(&snapshot_id, manifest))
        .await
        .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn skin_undo_abort(snapshot_id: String) -> Result<(), String> {
    skin_presets::undo_abort(&snapshot_id)
}

#[tauri::command]
pub async fn skin_catalog_fetch(
    source_id: String,
    force: Option<bool>,
) -> Result<Vec<skin_catalog::CatalogPack>, String> {
    skin_catalog::fetch_catalog(&source_id, force.unwrap_or(false)).await
}

#[tauri::command]
pub async fn skin_catalog_download(
    source_id: String,
    pack_id: String,
) -> Result<SkinPackPreviewDto, String> {
    skin_catalog::download_pack(&source_id, &pack_id).await
}

#[tauri::command]
pub async fn skin_catalog_preview_path(
    source_id: String,
    pack_id: String,
) -> Result<Option<String>, String> {
    skin_catalog::preview_path(&source_id, &pack_id).await
}

#[tauri::command]
pub async fn skin_sources_list() -> Result<Vec<skin_catalog::SkinSource>, String> {
    skin_catalog::list_sources()
}

#[tauri::command]
pub async fn skin_sources_add(url: String, label: Option<String>) -> Result<skin_catalog::SkinSource, String> {
    skin_catalog::add_source(&url, label.as_deref().unwrap_or(""))
}

#[tauri::command]
pub async fn skin_sources_remove(id: String) -> Result<(), String> {
    skin_catalog::remove_source(&id)
}

#[tauri::command]
pub async fn skin_sources_set_enabled(
    id: String,
    enabled: bool,
) -> Result<skin_catalog::SkinSource, String> {
    skin_catalog::set_source_enabled(&id, enabled)
}

#[tauri::command]
pub async fn skin_import_take_pending(
    slot: State<'_, Arc<PendingSlot>>,
) -> Result<Option<PendingSkinImport>, String> {
    Ok(skin_deeplink::take_pending(&slot))
}

#[tauri::command]
pub async fn skin_pack_fetch_url(href: String) -> Result<SkinPackPreviewDto, String> {
    skin_catalog::fetch_url_pack(&href).await
}
